processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+      final int maxBufferGrouperSize
+  )
+  {
+    this.buffer = buffer;
+    this.keySerde = keySerde;
+    this.keySize = keySerde.keySize();
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+    this.maxBufferGrouperSize = maxBufferGrouperSize;
+
+    int offset = HASH_SIZE + keySize;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+
+    this.bucketSize = offset;
+    this.tableArenaSize = buffer.capacity() / (bucketSize + Ints.BYTES) * bucketSize;
+
+    reset();
+  }
+
+  @Override
+  public boolean aggregate(KeyType key, int keyHash)
+  {
+    final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+    if (keyBuffer == null) {
+      return false;
+    }
+
+    Preconditions.checkArgument(
+        keyBuffer.remaining() == keySize,
+        "key size[%s] != keySize[%s]",
+        keyBuffer.remaining(),
+        keySize
+    );
+
+    int bucket = findBucket(
+        tableBuffer,
+        buckets,
+        bucketSize,
+        size < Math.min(maxSize, maxBufferGrouperSize),
+        keyBuffer,
+        keySize,
+        keyHash
+    );
+
+    if (bucket < 0) {
+      if (size < maxBufferGrouperSize) {
+        growIfPossible();
+        bucket = findBucket(tableBuffer, buckets, bucketSize, size < maxSize, keyBuffer, keySize, keyHash);
+      }
+
+      if (bucket < 0) {
+        return false;
+      }
+    }
+
+    final int offset = bucket * bucketSize;
+
+    // Set up key if this is a new bucket.
+    if (!isUsed(bucket)) {
+      tableBuffer.position(offset);
+      tableBuffer.putInt(keyHash | 0x80000000);
+      tableBuffer.put(keyBuffer);
+
+      for (int i = 0; i < aggregators.length; i++) {
+        aggregators[i].init(tableBuffer, offset + aggregatorOffsets[i]);
+      }
+
+      buffer.putInt(tableArenaSize + size * Ints.BYTES, offset);
+      size++;
+    }
+
+    // Aggregate the current row.
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].aggregate(tableBuffer, offset + aggregatorOffsets[i]);
+    }
+
+    return true;
+  }
+
+  @Override
+  public boolean aggregate(final KeyType key)
+  {
+    return aggregate(key, Groupers.hash(key));
+  }
+
+  @Override
+  public void reset()
+  {
+    size = 0;
+    buckets = Math.min(tableArenaSize / bucketSize, INITIAL_BUCKETS);
+
+    if (buckets < 1) {
+      throw new IAE(
+          "Not enough capacity for even one row! Need[%,d] but have[%,d].",
+          bucketSize + Ints.BYTES,
+          buffer.capacity()
+      );
+    }
+
+    maxSize = Math.max(1, (int) (buckets * MAX_LOAD_FACTOR));
+
+    // Clear used bits of new table
+    for (int i = 0; i < buckets; i++) {
+      buffer.put(i * bucketSize, (byte) 0);
+    }
```

Comment:
thanks for checking it... I'm concerned about the growth... each grow operation requires zero byte initialization, full copy of data and does not reuse old space... in the cases where multiple "grow" happen, performance will be poor.
it might make sense to make INITIAL_BUCKETS configurable at query level, so that users can choose it knowing their bucketSize... if bucketSize is in the order of KBs or more (which happens very commonly when sketches are used) then initialization cost might not be that significant.
Also, default value 32 is very small... we can easily start with something like 8192 .